llvm/lib/Target/Mips/NanoMipsOptimizeJumpTables.cpp

```diff
@@ -0,0 +1,289 @@
+    AU.addRequired<ProfileSummaryInfoWrapperPass>();
+    AU.addRequired<MachineLoopInfo>();
+    MachineFunctionPass::getAnalysisUsage(AU);
+  }
+};
+} // namespace
+
+INITIALIZE_PASS_BEGIN(NMOptimizeJumpTables, "nm-optimize-jt", NM_OPTIMIZE_JUMP_TABLES_OPT_NAME,
+                      false, false)
+INITIALIZE_PASS_DEPENDENCY(MachineBlockFrequencyInfo)
+INITIALIZE_PASS_DEPENDENCY(MachineBranchProbabilityInfo)
+INITIALIZE_PASS_DEPENDENCY(ProfileSummaryInfoWrapperPass)
+INITIALIZE_PASS_DEPENDENCY(MachineLoopInfo)
+INITIALIZE_PASS_END(NMOptimizeJumpTables, "nm-optimize-jt", NM_OPTIMIZE_JUMP_TABLES_OPT_NAME,
+                    false, false)
+
+char NMOptimizeJumpTables::ID = 0;
+
+// TODO: Currently, there is no existing LLVM interface which we can use to tell the
+// maximum potential size of a MachineInstr. Once we have it, this should be
+// enhanced.
+int NMOptimizeJumpTables::computeBlockSize(MachineBasicBlock &MBB) {
+  int Size = 0;
+  for (const MachineInstr &MI : MBB)
+    Size += TII->getInstSizeInBytes(MI);
+  return Size;
+}
+
+void NMOptimizeJumpTables::scanFunction() {
+  BlockInfo.clear();
+  BlockInfo.resize(MF->getNumBlockIDs());
+  BrOffsets.clear();
+  bool findBR = MF->getJumpTableInfo() &&
+                !MF->getJumpTableInfo()->getJumpTables().empty();
+  if (findBR)
+    BrOffsets.resize(MF->getJumpTableInfo()->getJumpTables().size());
+  int Offset = 0;
+  for (MachineBasicBlock &MBB : *MF) {
+    BlockInfo[MBB.getNumber()] = Offset;
+    Offset += computeBlockSize(MBB);
+    if (findBR)
+      for (auto &MI : MBB) {
+        if (MI.getOpcode() == Mips::BRSC_NM) {
+          int JTIdx = MI.getOperand(1).getIndex();
+          BrOffsets[JTIdx] = Offset;
+          break;
+        }
+      }
+  }
+}
+
+bool NMOptimizeJumpTables::optimizeRedundantEntries(
+    MachineBasicBlock::iterator &I) {
+  auto JTOp = (*I).getOperand(3);
+  int JTIdx = JTOp.getIndex();
+  auto &JTInfo = *MF->getJumpTableInfo();
+  const MachineJumpTableEntry &JT = JTInfo.getJumpTables()[JTIdx];
+  llvm::SmallPtrSet<MachineBasicBlock *, 16> JTMBBS;
+
+  // Collect all different JT MBBs.
+  for (auto MBB : JT.MBBs)
+    JTMBBS.insert(MBB);
+
+  MachineBasicBlock *MBBToJumpTo = nullptr;
+  MachineBasicBlock *CurrBB = JTOp.getParent()->getParent();
+
+  for (auto *MBB : JTMBBS) {
+    auto I = MBB->getFirstNonDebugInstr();
+    if (I != MBB->end() && !I->isUnconditionalBranch()) {
+      // Nonempty block.
+      if (!MBBToJumpTo) {
+        MBBToJumpTo = MBB;
+        continue;
+      }
+      // If this is the second nonempty, we can't optimize JT.
+      if (MBBToJumpTo != MBB)
+        return false;
+    }
+    // Empty block, or block with only unconditional jump.
+    // Check if every predeccessor is JT MBB, or CurrBB.
+    for (auto P : MBB->predecessors())
+      if (P != CurrBB && !JTMBBS.contains(P))
+        return false;
+
+    // Check if the block's successor is in JT MBBs. If it's not, it is a
```

Comment:
(Also, sorry if my initial comment wasn't clear in intent!)

Reply:
Regarding the current code, I wanted to avoid iterating multiple times over the same block. For instance, if we have empty blocks 2, 3, 4, and 5 that are members of JT, and let's say that 3 is a single successor of 2, etc.

> 
> ```c++
> for (auto *MBBi : JTMBBS) {
>   auto I = MBB->getFirstNonDebugInstr();
>   while (I == MBB->end() || I->isUnconditionalBranch()) {
>     // Empty block. Forward
>     MBB = MBB->getSingleSuccessor();
>     I = MBB->getFirstNonDebugInstr();
>   } 
>   if (!MBBToJumpTo)
>     MBBToJumpTo = MBB;
>   if (MBBToJumpTo != MBB)
>     return false;
> }
> ``
In this way, we would be iterating over MBB 5 when checking blocks  2, 3, 4, and 5. That's why I chose to stay "inside" the JT and do only one level of iteration(but still catch chains of empty blocks as long as they are all members of JT).

However, now that I see the code you suggested, it is much simpler to understand and also covers the case when we have a chain of empty blocks outside of JT, which is not covered by the current code. So, I agree with you that we should stick to the code you've suggested here. I will make the necessary changes. Thank you!